File: src/page/support/InListTH.java
package page.support;

import common.system.Copable;

import javax.swing.*;
import java.awt.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.awt.dnd.DragSource;
import java.io.IOException;

public class InListTH<T> extends TransferHandler implements Transferable {

    private static final long serialVersionUID = 1L;

    private final DataFlavor adf;
    private final ReorderList<T> list;

    private int obj;
    private int editing = -1;

    public InListTH(ReorderList<T> jlist) {
        list = jlist;
        adf = new DataFlavor(Integer.class, "");
        list.setDragEnabled(true);
        list.setDropMode(DropMode.INSERT);
    }

    public InListTH(ReorderList<T> jlist, Class<?> cls, String fla) {
        list = jlist;
        adf = new DataFlavor(cls, fla);
        list.setDragEnabled(true);
        list.setDropMode(DropMode.INSERT);
    }

    @Override
    public boolean canImport(TransferSupport info) {
        boolean b = info.getComponent() == list;
        b &= info.isDrop();
        b &= info.getDataFlavors()[0] == adf;
        list.setCursor(b ? DragSource.DefaultMoveDrop : DragSource.DefaultMoveNoDrop);
        return b;
    }

    @Override
    public int getSourceActions(JComponent c) {
        return TransferHandler.COPY_OR_MOVE;
    }

    @Override
    public T getTransferData(DataFlavor arg0) throws UnsupportedFlavorException, IOException {
        return list.copymap.get(obj);
    }

    @Override
    public DataFlavor[] getTransferDataFlavors() {
        return new DataFlavor[]{adf};
    }

    @SuppressWarnings("unchecked")
    @Override
    public boolean importData(TransferSupport info) {
        ReorderList<T> target = (ReorderList<T>) info.getComponent();
        assert target == list;
        if (info.isDrop()) {
            JList.DropLocation dl = (JList.DropLocation) info.getDropLocation();
            int fin = dl.getIndex();
            int max = list.getModel().getSize();
            if (fin < 0 || fin > max)
                fin = max;
            target.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
            try {
                int ori = editing;
                if (ori != -1 && ori != fin) {
                    list.reorder(ori, fin);
                    if (fin > ori)
                        fin--;
                    target.setSelectedIndex(fin);
                    return true;
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        } else {
            if (!list.copable)
                return false;
            Copable<T> t = (Copable<T>) list.copymap.get(obj);
            if (t == null)
                return false;
            return list.add(t.copy());
        }
        return false;
    }

    @Override
    public boolean isDataFlavorSupported(DataFlavor arg0) {
        return arg0.getHumanPresentableName().equals(adf.getHumanPresentableName());
    }

    @Override
    protected Transferable createTransferable(JComponent c) {
        assert (c == list);
        T t = list.getSelectedValue();
        if (list.copable)
            list.copymap.put(obj = t.hashCode(), t);
        editing = list.getSelectedIndex();
        return this;
    }

    @Override
    protected void exportDone(JComponent c, Transferable t, int act) {
        if (act == TransferHandler.MOVE || act == TransferHandler.NONE)
            list.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
    }

}

File: src/page/support/InListTH.kt
package page.support

import common.system.Copable
import java.awt.Cursor
import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.Transferable
import java.awt.datatransfer.UnsupportedFlavorException
import java.awt.dnd.DragSource
import java.io.IOException
import javax.swing.DropMode
import javax.swing.JComponent
import javax.swing.JList
import javax.swing.TransferHandler

class InListTH<T> : TransferHandler, Transferable {
    private val adf: DataFlavor
    private val list: ReorderList<T>
    private var obj = 0
    private var editing = -1

    constructor(jlist: ReorderList<T>) {
        list = jlist
        adf = DataFlavor(Int::class.java, "")
        list.setDragEnabled(true)
        list.setDropMode(DropMode.INSERT)
    }

    constructor(jlist: ReorderList<T>, cls: Class<*>?, fla: String?) {
        list = jlist
        adf = DataFlavor(cls, fla)
        list.setDragEnabled(true)
        list.setDropMode(DropMode.INSERT)
    }

    override fun canImport(info: TransferSupport): Boolean {
        var b = info.getComponent() === list
        b = b and info.isDrop()
        b = b and (info.getDataFlavors().get(0) === adf)
        list.setCursor(if (b) DragSource.DefaultMoveDrop else DragSource.DefaultMoveNoDrop)
        return b
    }

    override fun getSourceActions(c: JComponent): Int {
        return TransferHandler.COPY_OR_MOVE
    }

    @Throws(UnsupportedFlavorException::class, IOException::class)
    override fun getTransferData(arg0: DataFlavor): T {
        return list.copymap.get(obj)
    }

    override fun getTransferDataFlavors(): Array<DataFlavor> {
        return arrayOf<DataFlavor>(adf)
    }

    override fun importData(info: TransferSupport): Boolean {
        val target: ReorderList<T> = info.getComponent()
        assert(target === list)
        if (info.isDrop()) {
            val dl: JList.DropLocation = info.getDropLocation() as JList.DropLocation
            var fin: Int = dl.getIndex()
            val max: Int = list.getModel().getSize()
            if (fin < 0 || fin > max) fin = max
            target.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR))
            try {
                val ori = editing
                if (ori != -1 && ori != fin) {
                    list.reorder(ori, fin)
                    if (fin > ori) fin--
                    target.setSelectedIndex(fin)
                    return true
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        } else {
            if (!list.copable) return false
            val t: Copable<T> = list.copymap.get(obj) as Copable<T> ?: return false
            return list.add(t.copy())
        }
        return false
    }

    override fun isDataFlavorSupported(arg0: DataFlavor): Boolean {
        return arg0.getHumanPresentableName() == adf.getHumanPresentableName()
    }

    protected override fun createTransferable(c: JComponent): Transferable {
        assert(c === list)
        val t: T = list.getSelectedValue()
        if (list.copable) list.copymap.put(t.hashCode().also { obj = it }, t)
        editing = list.getSelectedIndex()
        return this
    }

    protected override fun exportDone(c: JComponent, t: Transferable, act: Int) {
        if (act == TransferHandler.MOVE || act == TransferHandler.NONE) list.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR))
    }

    companion object {
        private const val serialVersionUID = 1L
    }
}
